Add more log options

joaquimrocha opened the pull request:
This PR adds a few more options to the pod's log viewer.
Here is an illustrative screenshot:
![screenshot showing that there are now more options in the log viewer](https://user-images.githubusercontent.com/1029635/171037895-d6d2f940-5dfd-42c4-93d6-fb136664282c.png)

I have also added a new option in the number of tailed lines which allows to show all lines since the container started.

yolossn commented on frontend/src/components/pod/Details.tsx:
```diff
@@ -116,6 +127,18 @@ function PodLogViewer(props: PodLogViewerProps) {
             ))}
           </Select>
         </FormControl>,
+        <FormControlLabel
```
The option should only be made available only if the pod has restarted atleast once, else it will fail.

yolossn commented:
It makes sense to add a filter based on time which kubectl supports with the --since argument. Also to filter based on container if the pod contains more than one container.

joaquimrocha commented:
I have updated this PR to make the "Show previous" switch as disabled unless the container has been restarted, and also added tooltips:
![screenshot of the enabled previous button](https://user-images.githubusercontent.com/1029635/171417168-7e87b83e-809a-4101-ad9d-3d1ae91b0312.png)

![screenshot of the disabled previous button](https://user-images.githubusercontent.com/1029635/171417339-b010ea31-16af-4ea3-88e2-3924cb3609f7.png)

illume reviewed (commented):
Nice improvements.

Left a few comments and I found a bug where it doesn't work turning off the timestamps where lines "all" is selected. Here's a screenshot...
![show-time-stamps](https://user-images.githubusercontent.com/9541/171657364-a730532c-dd7e-45ea-91f2-52ddb881d54c.png)

joaquimrocha commented on frontend/src/lib/k8s/pod.ts:
```diff
@@ -31,6 +31,27 @@ export interface ExecOptions {
 }
 
+interface LogOptions {
+  tailLines?: number;
```
Good point about consistency among vars.
About the interface fields' docs. We don't usually add those to the things that correspond to the kubernetes types.
@yolossn has proposed that we generate these types, so maybe there's no real point in doing the effort of adding docs to the existing types right now.

joaquimrocha replied:
Ah, sorry. I confused these with the k8s types. Let me add the docs.

joaquimrocha commented on frontend/src/lib/k8s/pod.ts:
```diff
@@ -42,9 +63,22 @@ class Pod extends makeKubeObject<KubePod>('Pod') {
+  getLogs(...args: Parameters<oldGetLogs | newGetLogs>): () => void {
+    if (args.length > 3) {
+      console.debug(
+        "This Pod's getLogs use will soon be deprecated! Please double check how to call the getLogs function."
```
Makes sense to me. I think we should add the deprecation to the headlamp-plugin release notes, but I am not sure exactly what you mean by changelog.

joaquimrocha commented on frontend/src/components/pod/Details.tsx:
```diff
@@ -108,14 +140,52 @@ function PodLogViewer(props: PodLogViewerProps) {
+          <FormControlLabel
+            className={classes.switchControl}
+            label={t('logs|Show previous')}
+            disabled={!hasContainerRestarted()}
```
We do have a tooltip now describing it.

joaquimrocha commented on frontend/src/components/pod/Details.tsx:
```diff
@@ -108,14 +140,52 @@ function PodLogViewer(props: PodLogViewerProps) {
               </MenuItem>
             ))}
+            <MenuItem value="0" key="all">
+              {t('All')}
```
It has not... Maybe we just remove the option for now.

joaquimrocha commented on frontend/src/lib/k8s/pod.ts:
```diff
@@ -42,9 +63,22 @@ class Pod extends makeKubeObject<KubePod>('Pod') {
+      );
+      const [container, tailLines, showPrevious, onLogs] = args;
+      return this.getLogs(container, onLogs!, {
+        tailLines: tailLines as number,
```
I will have to add `as` to the args: ` args as Parameters<oldGetLogs>;` but I think it's better

joaquimrocha commented on frontend/src/components/pod/Details.tsx:
```diff
@@ -108,14 +140,52 @@ function PodLogViewer(props: PodLogViewerProps) {
             value={lines}
             onChange={handleLinesChange}
+            displayEmpty
+            renderValue={lines => (parseInt(lines as string) > 0 ? (lines as string) : t('All'))}
```
I will remove this "All" lines commit.

joaquimrocha commented:
@illume , I have updated the PR after your comments. Let me know if I tackled everything that makes sense.

illume reviewed (approved):
Thanks!
